controlplane/src/core/repositories/OrganizationRepository.ts

```diff
@@ -812,8 +816,46 @@ export class OrganizationRepository {
     return result[0];
   }
 
-  public deleteOrganization(organizationID: string) {
-    return this.db.delete(organizations).where(eq(organizations.id, organizationID)).execute();
+  public deleteOrganization(
+    organizationId: string,
+    organizationSlug: string,
+    opts: {
+      keycloakClient: Keycloak;
+      keycloakRealm: string;
+    },
+  ) {
+    return this.db.transaction(async (tx) => {
+      // Cancel subscription
+      const billingRepo = new BillingRepository(tx);
+      const billingService = new BillingService(tx, billingRepo);
+      const oidcRepo = new OidcRepository(tx);
+      const oidcProvider = new OidcProvider();
+
+      const subscription = await billingRepo.getActiveSubscriptionOfOrganization(organizationId);
+      if (subscription) {
+        await billingService.cancelSubscription(organizationId, subscription.id, 'Deleted by api');
```

Comment:
We should never do long-running or external work in a transaction. Secondly, in a rollback we would not rollback stripe so we can't rely on transactional rollback. I think we only need a tx for the last two statements. Therefore, the whole implementation needs to be implemented idempotent. Imagine, what should happen when `deleteOidcProvider` fails? It must be possible to rerun the method or more precisely the whole `deleteUser` procedure.

Reply:
improved